### PR TITLE
Add condvar implementation

### PR DESCRIPTION
**Pull request by tsoutsman**
Depends on https://github.com/theseus-os/Theseus/pull/942.

Signed-off-by: Klimenty Tsoutsman <klim@tsoutsman.com>

### CODE REVIEW

**Comment by tsoutsman** [comment]
Now this PR only depends on #942.

**Comment by kevinaboos** [comment]
@tsoutsman #942 has been merged in. thanks!

File: kernel/sync_block/src/condvar.rs [hunk 1]
```diff
@@ -0,0 +1,107 @@
+use crate::MutexGuard;
+use cpu_local_preemption::hold_preemption;
+use mpmc_queue::Queue;
+use sync::DeadlockPrevention;
+use sync_spin::Spin;
+use task::{get_my_current_task, TaskRef};
+
+/// A condition variable.
+// TODO: Is there even a point to exposing this generic?
```
Comment (tsoutsman) [inline kernel/sync_block/src/condvar.rs:12]: Bump.

**Comment by tsoutsman** [comment]
Should be good to go aside from the one comment I've left.

File: kernel/sync_block/src/condvar.rs (hunk 1 above)
Comment (kevinaboos) [reply kernel/sync_block/src/condvar.rs:12]: see discord #sync-primitives discussion, but yea, we can remove it.

File: kernel/sync_block/src/condvar.rs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    /// protect against spurious wakeups.
+    pub fn wait<'a, T>(&self, guard: MutexGuard<'a, T>) -> MutexGuard<'a, T> {
+        let task = get_my_current_task().unwrap();
+        let mutex = guard.mutex();
+
+        let preemption_guard = self
+            .inner
+            .push_if_fail(task.clone(), || {
+                drop(guard);
+                let preemption_guard = hold_preemption();
```
Comment (kevinaboos) [inline kernel/sync_block/src/condvar.rs]: this and other calls to `hold_preemption()` are likely good candidates for using `hold_preemption_no_timer_disable()` instead. lmk if you disagree.

**Review by kevinaboos — COMMENTED** [review]
Generally it looks pretty good. Will approve pending some discussion, one point of which is...

Could we potentially simplify the implementation using a waitqueue impl based on this new set of sync primitives? I only ask because it seems like there's a fair amount of overlap between the inner impl of Condvar (in the way that it uses the mpmc_queue) and what a Waitqueue offers.

File: kernel/sync_block/src/condvar.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    /// Blocks the current thread until this condition variable receives a
+    /// notification.
+    ///
+    /// Note that this function is susceptible to spurious wakeups. Condition
+    /// variables normally have a boolean predicate associated with them, and
+    /// the predicate must always be checked each time this function returns to
+    /// protect against spurious wakeups.
+    pub fn wait<'a, T>(&self, guard: MutexGuard<'a, T>) -> MutexGuard<'a, T> {
```
Comment (kevinaboos) [inline kernel/sync_block/src/condvar.rs]: I think we can relax the implicit Sized bound on T here, as Mutexes allow for that too.
```suggestion
    pub fn wait<'a, T: ?Sized>(&self, guard: MutexGuard<'a, T>) -> MutexGuard<'a, T> {
```

**Comment by tsoutsman** [comment]
> Could we potentially simplify the implementation using a waitqueue impl

No I don't think so, because we perform a different action on the first `push_if_fail` and a wait queue doesn't allow for that.

File: kernel/sync_block/src/condvar.rs (hunk 3 above)
Comment (tsoutsman) [reply kernel/sync_block/src/condvar.rs]: Depends on #962.

Comment (kevinaboos) [reply kernel/sync_block/src/condvar.rs]: merged in #962, thanks.

**Comment by tsoutsman** [comment]
@kevinaboos should be good to merge now

**Review by kevinaboos — APPROVED** [review]

**Comment by kevinaboos** [comment]
> @kevinaboos should be good to merge now

Thanks, approved. If you want to remove that "TODO" item about the `P = Spin` parameter on `Condvar`, feel free to follow-up with another PR. I'm ok with it being there or not, either is fine.

**Pull request merged**